cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,699 @@
+    if (!keys.empty()) {
+      metadata = std::make_shared<KeyValueMetadata>();
+      for (auto it = keys.begin(); it != keys.end(); ++it) {
+        metadata->Append(*it, reader_->getMetadataValue(*it));
+      }
+    }
+
+    *out = std::make_shared<Schema>(fields, metadata);
+    return Status::OK();
+  }
+
+  Status Read(std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    return read_batch(opts, NumberOfRows(), out);
+  }
+
+  Status Read(const std::list<uint64_t>& include_indices,
+              std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    opts.includeTypes(include_indices);
+    return read_batch(opts, NumberOfRows(), out);
+  }
+
+  Status ReadStripe(int64_t stripe, std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    RETURN_NOT_OK(select_stripe(&opts, stripe));
+    return read_batch(opts, stripes_[stripe].num_rows, out);
+  }
+
+  Status ReadStripe(int64_t stripe, const std::list<uint64_t>& include_indices,
+                    std::shared_ptr<RecordBatch>* out) {
+    liborc::RowReaderOptions opts;
+    RETURN_NOT_OK(select_stripe(&opts, stripe));
+    opts.includeTypes(include_indices);
+    return read_batch(opts, stripes_[stripe].num_rows, out);
+  }
+
+  Status select_stripe(liborc::RowReaderOptions* opts, int64_t stripe) {
+    if (stripe < 0 || stripe >= NumberOfStripes()) {
+      std::stringstream ss;
+      ss << "Out of bounds stripe: " << stripe;
+      return Status::Invalid(ss.str());
+    }
+    opts->range(stripes_[stripe].offset, stripes_[stripe].length);
+    return Status::OK();
+  }
+
+  Status read_batch(const liborc::RowReaderOptions& opts, int64_t nrows,
+                    std::shared_ptr<RecordBatch>* out) {
+    std::unique_ptr<liborc::RowReader> rowreader;
+    std::unique_ptr<liborc::ColumnVectorBatch> batch;
+    try {
+      rowreader = reader_->createRowReader(opts);
+      batch = rowreader->createRowBatch(std::min(nrows, kReadRowsBatch));
+    } catch (const liborc::ParseError& e) {
+      return Status::Invalid(e.what());
+    }
+    const liborc::Type& type = rowreader->getSelectedType();
+    std::shared_ptr<Schema> schema;
+    RETURN_NOT_OK(type_to_schema(type, &schema));
+
+    std::unique_ptr<RecordBatchBuilder> builder;
+    RETURN_NOT_OK(RecordBatchBuilder::Make(schema, pool_, nrows, &builder));
+
+    // The top-level type must be a struct to read into an arrow table
+    auto struct_batch = static_cast<liborc::StructVectorBatch*>(batch.get());
+
+    int n_fields = builder->num_fields();
+    while (rowreader->next(*batch)) {
+      for (int i = 0; i < n_fields; i++) {
+        RETURN_NOT_OK(append_batch(type.getSubtype(i), builder->GetField(i),
+                                   struct_batch->fields[i], 0, batch->numElements));
+      }
+    }
+    RETURN_NOT_OK(builder->Flush(out));
+    return Status::OK();
+  }
+
+  Status append_batch(const liborc::Type* type, ArrayBuilder* builder,
```

Comment:
`AppendBatch`. Also, can we change the argument orders and some refs/pointers here -- "in" arguments should come first, followed by the in-out argument `builder`. `batch` should be passed as const-ref (initially I thought that `type` should be too, but I see that the ORC API returns pointers to type objects)

Reply:
The ORC api also returns pointers for `ColumnVectorBatch` objects, so I've left those as pointers as well.